docs/deploying.adoc

```diff
@@ -9,11 +9,250 @@ Put the filter jars in some directory
+alternatives are supported: `PortPerBroker` and `SniRouting` which have different characteristics which make each
+suitable for different use-cases. They are described next.
+
+==== PortPerBroker scheme
+
+In the `PortPerBroker` scheme, Kroxylicious automatically opens a port for each virtual cluster's bootstrap and
+one port per broker of each target cluster.  So, if you have two virtual clusters, each targeting a Kafka Cluster
+of three brokers, Kroxylicious will bind eight ports in total.
+
+Kroxylicious monitors the broker topology of the target cluster at runtime. It will adjust the number of open ports
+dynamically.  If another broker is added to the Kafka Cluster, Kroxylicious will automatically open a port for it.
+Similarly, if a broker is removed from the Kafka Cluster, Kroxylicious will automatically close the port that was
+assigned to it.
+
+The `PortPerBroker` scheme can be used with either plain or TLS downstream connections.
+
+[source, yaml]
+----
+clusterNetworkAddressConfigProvider:
+  type: PortPerBroker
+  config:
+    bootstrapAddress: mycluster.kafka.com:9192                                 # <1>
+    brokerAddressPattern: mybroker-$(nodeId).mycluster.kafka.com:$(portNumber) # <2>
+    brokerStartPort: 9193                                                      # <3>
+    numberOfBrokerPorts: 3                                                     # <4>
+    bindAddress: 192.168.0.1                                                   # <5>
+----
+<1> The hostname and port of the bootstrap that will be used by the Kafka Clients.
+<2> (Optional) The broker address pattern used to form the broker addresses.  If not defined, it defaults to the
+hostname part of the `bootstrapAddress` and the port number allocated to the broker.
+<3> (Optional) The starting number for broker port range. Defaults to the port of the `bootstrapAddress` plus 1.
+<4> (Optional) The _maximum_ number of brokers of ports that are permitted.  Defaults to 3.
+<5> (Optional) The bind address used when binding the ports. If undefined, all network interfaces will be bound.
+
+The `brokerAddressPattern` configuration parameters understands two replacement tokens.
+
+* this `$(portNumber)` is mandatory.  It is replaced by the port number computed for this broker.
+* the `$(nodeId)` token is optional.  If present, it will be replaced by the
+https://kafka.apache.org/documentation/#brokerconfigs_node.id[`node.id`] (or `broker.id`) assigned to the broker
+of the target cluster.
+
+For example if your configuration looks like the above and your cluster has three brokers, your Kafka Client will receive
+broker address information like this:
+
+0.  `mybroker-0.mycluster.kafka.com:9193`
+1.  `mybroker-1.mycluster.kafka.com:9194`
+2.  `mybroker-2.mycluster.kafka.com:9194`
+
+NOTE: It is a responsible for the deployer of Kroxylicious to ensure that the bootstrap and generated broker
+DNS names are resolvable and routable by the Kafka Client.
```

Comment:
Depends how richly we want to support bare metal deployments really as there is not going to be an operator there. 

An API hook also possibly leads to race conditions between kroxylicious instances which might need some co-ordination. 

A second use case for the API hook would be kroxylicious registering itself as a back end server for an external load balancer (e.g. its got the metadata so now it can join the pool of servers backing the load balancer).